core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -211,32 +230,49 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
         rawConnectivityState = IDLE;
         updateBalancingState(IDLE, new RequestConnectionPicker(this));
         break;
+
       case CONNECTING:
         rawConnectivityState = CONNECTING;
         updateBalancingState(CONNECTING, new Picker(PickResult.withNoResult()));
         break;
+
       case READY:
         shutdownRemaining(subchannelData);
         addressIndex.seekTo(getAddress(subchannel));
         rawConnectivityState = READY;
         updateHealthCheckedState(subchannelData);
         break;
+
       case TRANSIENT_FAILURE:
         // If we are looking at current channel, request a connection if possible
         if (addressIndex.isValid()
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
+          if (addressIndex.increment()) {
+            cancelScheduleTask();
+            requestConnection(); // is recursive so might hit the end of the addresses
+          }
+        }
 
-          // If no addresses remaining, go into TRANSIENT_FAILURE
-          if (!addressIndex.isValid()) {
+        // If we are (at the end and have had enough TF since we got there) or
+        // (just got to the end), then refresh name resolution which may call
+        // acceptResolvedAddresses which will reset addressIndex, but may not.
+        if (!addressIndex.isValid()) {
+          numTf.getAndIncrement();
```

Comment:
It looks like `numTF` doesn't need to be atomic. And that's good, because these three calls to adjust it wouldn't be thread-safe. We should make it a normal integer.

Reply:
Changed to a normal int.